third_party/nvfuser/csrc/scheduler/normalization_utils.cpp

```diff
@@ -494,5 +495,133 @@ std::optional<GridOuterNormalizationParams> getGridOuterNormalizationParams(
   return std::nullopt;
 }
 
+bool checkIfReductionsAreInnerOuter(
+    const std::vector<TensorView*>& inner_reduction_tvs,
+    const std::vector<TensorView*>& outer_reduction_tvs) {
+  bool pass_combined_heck = true;
+  // inner reduction must be [I,I,...R,R]
+  auto innerReductionCheck = [](TensorView* tv) {
+    int ndim = static_cast<int>(tv->nDims());
+    int lastIter = -1;
+    while (lastIter < ndim - 1 && tv->axis(lastIter + 1)->isIteration()) {
+      lastIter++;
+    }
+    int firstRedu = ndim;
+    while (firstRedu > 0 && tv->axis(firstRedu - 1)->isReduction()) {
+      firstRedu--;
+    }
+    return lastIter >= 0 && firstRedu < ndim && lastIter == firstRedu - 1;
+  };
+  // outer reduction must be [R,R,..I,I]
+  auto outerReductionCheck = [](TensorView* tv) {
+    int ndim = static_cast<int>(tv->nDims());
+    int lastRedu = -1;
+    while (lastRedu < ndim - 1 && tv->axis(lastRedu + 1)->isReduction()) {
+      lastRedu++;
+    }
+    int firstIter = ndim;
+    while (firstIter > 0 && tv->axis(firstIter - 1)->isIteration()) {
+      firstIter--;
+    }
+    return lastRedu >= 0 && firstIter < ndim && lastRedu == firstIter - 1;
+  };
+  for (auto itv : inner_reduction_tvs) {
+    if (!innerReductionCheck(itv)) {
+      pass_combined_heck = false;
+      break;
+    }
+  }
+  for (auto otv : outer_reduction_tvs) {
+    if (!outerReductionCheck(otv)) {
+      pass_combined_heck = false;
+      break;
+    }
+  }
+  return pass_combined_heck;
+}
+
+bool hasSharedInput(
+    const std::vector<TensorView*>& inner_reduction_tvs,
+    const std::vector<TensorView*>& outer_reduction_tvs) {
+  bool has_shared_input = false;
+  std::unordered_set<TensorView*> input_inner_reduction_tvs;
+  for (auto tv : inner_reduction_tvs) {
+    for (auto input_tv : ir_utils::inputTvsOf(tv)) {
+      input_inner_reduction_tvs.emplace(input_tv);
+    }
+  }
+  for (auto tv : outer_reduction_tvs) {
+    for (auto input_tv : ir_utils::inputTvsOf(tv)) {
+      if (input_inner_reduction_tvs.find(input_tv) !=
+          input_inner_reduction_tvs.end()) {
+        has_shared_input = true;
+        break;
+      }
+    }
+    if (has_shared_input) {
+      break;
+    }
+  }
+  return has_shared_input;
+}
+
+bool hasSharedConsumer(
+    const std::vector<TensorView*>& inner_reduction_tvs,
+    const std::vector<TensorView*>& outer_reduction_tvs) {
+  // check inner reduction tvs
+  auto contvs = ir_utils::consumerTvsOf(inner_reduction_tvs[0]);
```

Comment:
I think more fundamentally, we need to make sure all the tensors in a fusion are divided to two disjoint groups (assuming there's only one inner reduction and one outer reduction). One group should contain all tensors involved in the inner reduction and the tensors used for the first partial outer reduction. The another group should contain all tensors used for the final step of the outer reduction. We then need to make sure the only path between the two disjoint groups is the temp buffer, and only in that case the current way of limiting the propagations using the boundary node of the temp buffer makes sense.

Looking at the consumers of the inner reduction TVs doesn't seem to be sufficient, even if indirect consumers are considered. For example, what if there's a some tensor that's used by both in the inner and outer groups?

Reply:
I revised this fuction to grab all the consumers. It will check if outer reduction tvs have any shared consumer (1) with inner reduction tvs and (2) other outer reduction tvs.
I added an additional test case, `FusionCombinedSchedulerLinked_CUDA` where 
```
// This case is to test the correctness of the combined inner and outer
// scheduler, if link_inner_outer = true, the inner and outer reductions are
// linked, otherwise the two outer reductions are linked. In either case, the
// fusion should be segmented since the current combined scheduler assumes there
// is no shared consumer between inter reductions and outer reductions and among
// tensors in outer reductions.
```
If there is a tensor used by both the inner readuction and oute reduction, actually this is required by this combined scheduler, the partial outer reduction will do the appropriate usage of that tensor and dump the partial result to a tmp gmem. There should't be a tensor used by both the inner reduction and the final part of outer reduction.

Reply:
> If there is a tensor used by both the inner reduction and oute reduction, actually this is required by this combined scheduler,

I meant that one tensor is used for an inner reduction and the second phase of an outer reduction, i.e., it appears both before and after the grid sync. This tensor is not a consumer of any of the reduction tensors, so it wouldn't be flagged as an invalid pattern. However, the transform propagations could follow through the tensor, so the boundary-based approach doesn't seem to work.

Reply:
I don't think this pattern exist. If it is not a consumer of any of the reduciton tensors, it must be a producer of the reduction tensors. Whether it is a producer of outer reduction or inner reduction, it will be used in the first part to calculate the partial outer reduction or inner reduction. quote my previous reply:
```
There should't be a tensor used by both the inner reduction and the final part of outer reduction. All the usage is done when calculating the partial result of outer reduciton.
```

Reply:
I should have said that one tensor is used by the inner-reduction group and also the post-temp-buffer outer reduction group.